src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state }; // Create a clone of the initial state
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {}; // Clear the state
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key]; // Remove specified keys
+        });
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    stateHistory.push({ ...currentState }); // Save a clone of the current state
```

Comment:
The `stateHistory.push({ ...currentState });` statement should be moved outside of the `switch` block, but still within the `forEach` loop. This ensures that the state is pushed to the history after processing each action, as per the additional prompt rules.